### PR TITLE
RATIS-1927. Use double check to eliminate data race in ReadRequests 

### PR DESCRIPTION
**Pull request by SzyWilliam**
see https://issues.apache.org/jira/browse/RATIS-1927.

The `LastAppliedTermIndex` may be updated in between of the `check` and `enqueue`. 
Normally this data race is not a big deal since the next increment in AppliedIndex will complete the waiting entry. 
However, when IOTDB restarts, there will be no subsequent read/write requests unless the first probing read request successfully returns. Under situations like that, this data race may haunt the system several seconds until reaching timeout.

I added a post-double-check as a quick fix instead of implementing synchronization.

### CODE REVIEW

**Review by szetszwo — COMMENTED** [review]
@SzyWilliam , thanks for working on this!  Please see the comment inlined.

BTW, I recently found that atomic operations (e.g. `AtomicReference`/`ConcurrentMap`) required side-effect-free function; see RATIS-1923.  In `ReadIndexQueue.add(..)`, the `MemoizedSupplier` may be initialized a future but then discarded by `sorted.computeIfAbsent(..)` since it keeps retrying in a loop.  We may fixed it by creating the future in advance:
```java
//ReadIndexQueue
    CompletableFuture<Long> add(long readIndex) {
      final CompletableFuture<Long> created = new CompletableFuture<>();
      final CompletableFuture<Long> returned = sorted.computeIfAbsent(readIndex, i -> created);

      if (created == returned) {
        scheduler.onTimeout(readTimeout, () -> handleTimeout(readIndex),
            LOG, () -> "Failed to handle read timeout for index " + readIndex);
      }
      return returned;
    }
```

**Comment by szetszwo** [comment]
I wonder if it is more efficient to use `synchronized` to implement `ReadIndexQueue` than `ConcurrentSkipListMap`?
```java
  static class ReadIndexQueue {
    private final TimeoutExecutor scheduler = TimeoutExecutor.getInstance();
    private final NavigableMap<Long, CompletableFuture<Long>> sorted = new TreeMap<>();
    private final TimeDuration readTimeout;

    ReadIndexQueue(TimeDuration readTimeout) {
      this.readTimeout = readTimeout;
    }

    CompletableFuture<Long> add(long readIndex) {
      final CompletableFuture<Long> returned;
      final boolean create;
      synchronized (this) {
        // The same as computeIfAbsent except that it also tells if a new value is created.
        final CompletableFuture<Long> existing = sorted.get(readIndex);
        create = existing == null;
        if (create) {
          returned = new CompletableFuture<>();
          sorted.put(readIndex, returned);
        } else {
          returned = existing;
        }
      }

      if (create) {
        scheduler.onTimeout(readTimeout, () -> handleTimeout(readIndex),
            LOG, () -> "Failed to handle read timeout for index " + readIndex);
      }
      return returned;
    }

    private void handleTimeout(long readIndex) {
      final CompletableFuture<Long> removed;
      synchronized (this) {
        removed = sorted.remove(readIndex);
      }
      if (removed == null) {
        return;
      }
      removed.completeExceptionally(new ReadException("Read timeout " + readTimeout + " for index " + readIndex));
    }


    /** Complete all the entries less than or equal to the given applied index. */
    synchronized void complete(Long appliedIndex) {
      final NavigableMap<Long, CompletableFuture<Long>> headMap = sorted.headMap(appliedIndex, true);
      headMap.values().forEach(f -> f.complete(appliedIndex));
      headMap.clear();
    }
  }
```

**Comment by SzyWilliam** [comment]
> BTW, I recently found that atomic operations (e.g. AtomicReference/ConcurrentMap) required side-effect-free function; see [RATIS-1923](https://issues.apache.org/jira/browse/RATIS-1923). In ReadIndexQueue.add(..), the MemoizedSupplier may be initialized a future but then discarded by sorted.computeIfAbsent(..) since it keeps retrying in a loop. We may fixed it by creating the future in advance

Thanks for the explaination! Now I knows why we should use a side-effect-free function!

**Comment by SzyWilliam** [comment]
> I wonder if it is more efficient to use synchronized to implement ReadIndexQueue than ConcurrentSkipListMap?

Thanks very much for the patch! Indeed we should complete the future actively.

Additionally, I conducted an in-depth analysis of 'ConcurrentSkipListMap' and the synchronized 'TreeMap.' Both data structures have search, insert, and remove operations with a time complexity of O(logN). When removing multiple entries in `complete`, 'TreeMap' requires 2O(logN) + k operations (since it only needs to identify the ceiling and flooring entry), while 'ConcurrentSkipListMap' demands kO(logN) operations (as it loops over all the matching indexes in between). In terms of synchronization, synchronized 'TreeMap' uses locks, while 'ConcurrentSkipListMap' heavily relies on CAS operations with spinning.

In conclusion, synchronized 'TreeMap' demonstrates better performance under heavy loads and offers higher throughput, which is typical for Ratis read situations. On the other hand, 'ConcurrentSkipListMap' excels in terms of latency performance.

**Comment by SzyWilliam** [comment]
@szetszwo Applied corresponding changes on code. PTAL~

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java [hunk 1]
```diff
@@ -92,9 +99,15 @@ Consumer<Long> getAppliedIndexConsumer() {
...
   CompletableFuture<Long> waitToAdvance(long readIndex) {
-    if (stateMachine.getLastAppliedTermIndex().getIndex() >= readIndex) {
-      return CompletableFuture.completedFuture(readIndex);
+    final long lastApplied = stateMachine.getLastAppliedTermIndex().getIndex();
+    if (lastApplied >= readIndex) {
+      return CompletableFuture.completedFuture(lastApplied);
+    }
+    final CompletableFuture<Long> f = readIndexQueue.add(readIndex);
+    final long current = stateMachine.getLastAppliedTermIndex().getIndex();
+    if (current >= lastApplied) {
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java:108]: This should be `current > lastApplied`; otherwise, the condition is always true.

The change looks good other than that.

**Comment by szetszwo** [comment]
@SzyWilliam , thanks for researching the `ConcurrentSkipListMap` and the synchronized `TreeMap`!  I have learned a lot.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java (hunk 1 above)
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java:108]: You are right! I was thinking about `current >= readIndex`. Thanks very much for the careful reviews ;)

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by OneSizeFitsQuorum** [comment]
Hi, @szetszwo,  I checked the computeIfAbsent code in ConcurrentHashMap, it seems that the mapping function will be called just once, see issue for [details](https://issues.apache.org/jira/browse/RATIS-1923). So is it side-effect-free? If so, is it necessary to split a hash computation into two get & puts with a large lock
